### ISSUE TITLE
RyuJIT may violate the CLR memory model during CSE

### ISSUE THREAD

**Issue by pgavlin** [dotnet/coreclr#2286]
Test `JIT\Regression\CLR-x86-JIT\V1-M13-RTM\b99969` attempts to validate that the JIT does not improperly CSE array accesses across volatile memory operations. In particular, in the fragment below (which is a distillation of the aforementioned test), the second access of `arr[_index]` should not be CSE'd with the first, as doing so moves the read of `_index` above the read of `_flag`, thus violating the memory model.

```
sealed class C
{
    public int _index = 0;
    private volatile bool _flag = false;

    void M()
    {
        var arr = new int[1];

        int result = 0;
        result += arr[_index];

        while (!_flag);
        result += arr[_index];
    }
}
```

**Comment by briansull** [comment]
Pull request #2569 fixes this issue

**Comment by briansull** [comment]
Fix is checked in,
@pgavlin can you reenable this test
briansull merged commit dd0c2f1 into dotnet:master an hour ago

**Comment by pgavlin** [comment]
Closed by #2569.
